datatypes/ring_programs.py

```diff
@@ -776,10 +781,9 @@ def _time_from_turn(self, mass, circumference):
 
         """
         trev = rt.mom_to_trev(self[0], mass, circ=circumference)
-        return np.cumsum(trev)
-
+        return np.insert(np.cumsum(trev[:-1]), 0, 0)
```

Comment:
Is this needed for the `n_turns+1` thing?

Reply:
No that's to avoid getting a t_rev offset on the cycle_time (passing a program by_time would start from 0 or t_start while by_turn would start from t_rev)

Reply:
Ooh, I see.  Why not just do:
```python
t_rev -= t_rev[0]
```
?

Reply:
Definitely an option! I am wondering for very large arrays what is the most efficient. This can come at a second stage when we can evaluate how long does it take to generate a Ring object for our default use cases (although I believe this part won't be that critical)

Reply:
Definitely slower to use the insert, but not enough to matter I think.
```ipython
In [38]: test = np.linspace(0, 1, 10000)

In [39]: %timeit np.cumsum(test)
37.2 µs ± 58.6 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)

In [40]: %timeit np.cumsum(test) - 0.1
49 µs ± 146 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)

In [41]: %timeit np.insert(np.cumsum(test[:-1]), 0, 0)
82.7 µs ± 178 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
```

I think in this instance speed isn't really a concern, I'm just wondering if missing the last point could change the result

Reply:
Ah yes you reminded me, one thing we need to consider is indexing, to keep
cycle_time[i] = cycle_time[i-1] + t_rev[i-1] with i>0

Like this np.diff(cycle_time) == t_rev (with the last element missing) aligned with np.diff(energy) == delta_E

(I hope I didn't get the index wrong 😄)